storage/replica_command.go

```diff
@@ -473,6 +473,13 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, h ro
 	// Run triggers if successfully committed.
 	if reply.Txn.Status == roachpb.COMMITTED {
 		ct := args.InternalCommitTrigger
+		if ct != nil {
+			// Hold readMu across the application of any commit trigger.
+			// This makes sure that no reads are happening in parallel;
+			// see #3148.
+			r.readMu.Lock()
+			batch.Defer(r.readMu.Unlock)
```

Comment:
Some commit triggers also use batch.Defer; it might be worth a comment that these will run outside the scope of `readMu`.